lib/vnet/diag/routeconflict_darwin.go

```diff
@@ -0,0 +1,114 @@
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package diag
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/netip"
+	"os/exec"
+	"strings"
+	"syscall"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/net/route"
+)
+
+type DarwinRouting struct{}
+
+func (dr *DarwinRouting) GetRouteDestinations() ([]RouteDest, error) {
+	rib, err := route.FetchRIB(syscall.AF_INET, route.RIBTypeRoute, 0)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	messages, err := route.ParseRIB(route.RIBTypeRoute, rib)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	rds := make([]RouteDest, 0, len(messages))
+
+	for _, m := range messages {
+		rm, ok := m.(*route.RouteMessage)
+		if !ok {
+			return nil, trace.Errorf("expected *route.RouteMessage, got %T", m)
+		}
+		rd, err := routeMessageToDest(rm)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		rds = append(rds, rd)
+	}
+
+	return rds, nil
+}
+
+func routeMessageToDest(rm *route.RouteMessage) (RouteDest, error) {
+	rawDest := rm.Addrs[syscall.RTAX_DST]
+	destInet4Addr, ok := rawDest.(*route.Inet4Addr)
+	if !ok {
+		return nil, trace.Errorf("expected destination to be *route.Inet4Addr, got %T", rawDest)
+	}
+
+	var ip4Mask net.IPMask
+	destNetipAddr := netip.AddrFrom4(destInet4Addr.IP)
+	rawMask := rm.Addrs[syscall.RTAX_NETMASK]
+
+	// The route destination has no netmask, return just the IP address.
+	if rawMask == nil {
+		return &RouteDestIP{Addr: destNetipAddr, ifaceIndex: rm.Index}, nil
+	}
+
+	mask, ok := rawMask.(*route.Inet4Addr)
+	if !ok {
+		return nil, trace.Errorf("expected netmask to be *route.Inet4Addr, got %T", rawMask)
+	}
+	ip4Mask = net.IPv4Mask(mask.IP[0], mask.IP[1], mask.IP[2], mask.IP[3])
+
+	ones, _ := ip4Mask.Size()
+	prefix := netip.PrefixFrom(destNetipAddr, ones)
+	return &RouteDestPrefix{Prefix: prefix, ifaceIndex: rm.Index}, nil
+}
+
+// InterfaceApp returns desc field of NetworkExtension that's included in the output of `ifconfig -v
+// <ifaceName>` if the interface was created by an app that uses the NetworkExtension macOS
+// framework. Returns an empty string if no description was found.
+//
+// According to a post from Apple Developer Forums from 2019, that's the only way this description
+// can be extracted given an interface name.
+// https://developer.apple.com/forums/thread/113491
+func (n *NetInterfaces) InterfaceApp(ctx context.Context, ifaceName string) (string, error) {
```

Comment:
I got a comment from Alan to keep exported APIs in the platform-independent file so they're easier to find/reason about, and call to a platform-specific unexported function where necessary. Makes sense to me and I've been trying to follow that pattern